cmd/tke-installer/app/installer/expansion.go

```diff
@@ -0,0 +1,206 @@
+ * this file except in compliance with the License. You may obtain a copy of the
+ * License at
+ *
+ * https://opensource.org/licenses/Apache-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an “AS IS” BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package installer
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"strings"
+	"tkestack.io/tke/cmd/tke-installer/app/installer/constants"
+)
+
+// expansionPath is the dir where users put there expansion files. default is data/expansions/
+var expansionPath string
+
+// expansionEnabled
+var expansionEnabled bool
+
+const envExpansionPath = "EXPANSION_PATH"
+const expansionDirName = "expansions/"
+const defaultExpansionPath = constants.DataDir + expansionDirName
+
+// expansion layout
+const expansionLayoutChartsDir = "charts/"
+const expansionLayoutFilesDir = "files/"
+const expansionLayoutFilesGeneratedDir = "files_generated/"
+const expansionLayoutHooksDir = "hooks/"
+const expansionLayoutProviderDir = "provider/"
+const expansionLayoutImagesName = "images.tar.gz" //nolint
+const expansionLayoutConfDir = "conf/"
+const expansionLayoutApplicationsDir = "applications/"
+
+var expansionLayoutMap map[string]string
+
+// expansion resource
+const expansionResourceChart = "chart"
+const expansionResourceFile = "file"
+const expansionResourceFileGenerated = "file_generated"
+const expansionResourceHook = "hook"
+const expansionResourceProvider = "provider"
+const expansionResourceConf = "conf"
+const expansionResourceApplication = "application"
+
+// expansion value
+const expansionValuesFile = "values.yaml" //nolint
+
+func init() {
+
+	expansionPath = os.Getenv(envExpansionPath)
+	if expansionPath == "" {
+		expansionPath = defaultExpansionPath
+	}
+	if fi, err := os.Stat(expansionPath); err == nil {
+		if fi.IsDir() {
+			expansionEnabled = true
+		}
+	}
+
+	expansionLayoutMap = make(map[string]string)
+	if expansionEnabled {
+		expansionLayoutMap[expansionResourceChart] = path.Join(expansionPath, expansionLayoutChartsDir)
+		expansionLayoutMap[expansionResourceFile] = path.Join(expansionPath, expansionLayoutFilesDir)
+		expansionLayoutMap[expansionResourceFileGenerated] = path.Join(expansionPath, expansionLayoutFilesGeneratedDir)
+		expansionLayoutMap[expansionResourceHook] = path.Join(expansionPath, expansionLayoutHooksDir)
+		expansionLayoutMap[expansionResourceProvider] = path.Join(expansionPath, expansionLayoutProviderDir)
+		expansionLayoutMap[expansionResourceConf] = path.Join(expansionPath, expansionLayoutConfDir)
+		expansionLayoutMap[expansionResourceApplication] = path.Join(expansionPath, expansionLayoutApplicationsDir)
+	}
+}
+
+func (t *TKE) prepareExpansionFiles(ctx context.Context) error {
+
+	if !t.isExpansionEnabled() {
```

Comment:
If your step has a global switch, isExpansionEnabled, it should be out of your step